### PR TITLE
Reconstruct application specific data.

### PR DESCRIPTION
**Pull request by aterentic-ethernal**

### CODE REVIEW

File: kate/recovery/src/com.rs [hunk 1]
```diff
@@ -1,39 +1,115 @@
...
+				cells.insert(row, cell);
+				result.insert(col, cells);
+			},
+			Some(cells) => {
+				if cells.contains_key(&cell.row) {
+					return Err(anyhow!("Duplicate cell found"));
+				}
+				cells.insert(cell.row, cell);
+			},
+		}
```
Comment (MiguelDD1) [inline kate/recovery/src/com.rs:40]: `Entry` interface can simplify this.
```suggestion
	        let cells = result.entry(col).or_insert_with(HashMap::new);
		if cells.insert(cell.row, cell).is_some() {
			return Err(anyhow!("Duplicate cell found"));
		}
```
```

File: kate/recovery/src/com.rs [hunk 2]
```diff
@@ -1,39 +1,115 @@
...
+use anyhow::{anyhow, Result};
 use dusk_bytes::Serializable;
 use dusk_plonk::{fft::EvaluationDomain, prelude::BlsScalar};
 
 // TODO: Constants are copy from kate crate, we should move them to common place
 pub const DATA_CHUNK_SIZE: usize = 31;
 const PADDING_TAIL_VALUE: u8 = 0x80;
 
-// Reconstructs app extrinsics from extrinsics layout and data
+pub struct MatrixDimensions {
```
[the remark this answers is not in the document]
Comment (aterentic-ethernal) [reply kate/recovery/src/com.rs:11]: That should be one type, recovery crate cannot depend on kate crate :/ I'm looking atm how and where to put common data structures if it's possible without big changes.

File: kate/recovery/src/com.rs (hunk 1 above)
Comment (aterentic-ethernal) [reply kate/recovery/src/com.rs:40]: Very nice 🥇

File: kate/recovery/src/com.rs [hunk 3]
```diff
@@ -1,39 +1,106 @@
...
+	pub rows: usize,
+	pub cols: usize,
+	pub chunk_size: usize,
+}
+
+fn map_cells(
```
Comment (luka-ethernal) [inline kate/recovery/src/com.rs:17]: Just a quick docstring to describe what this fn does would be nice. Doesn't have to be more than one line.

File: kate/recovery/src/com.rs [hunk 4]
```diff
@@ -1,39 +1,106 @@
...
+		.collect::<Vec<_>>();
+	Ok(unflatten_padded_data(ranges, data, dimensions.chunk_size))
 }
 
 fn trim_to_chunk_data(chunk: &[u8]) -> [u8; DATA_CHUNK_SIZE] {
 	assert!(DATA_CHUNK_SIZE < chunk.len(), "Cannot trim to bigger size!");
 	chunk[0..DATA_CHUNK_SIZE].try_into().unwrap()
 }
 
+pub fn data_ranges(layout: &[(u32, u32)]) -> Vec<(u32, Range<usize>)> {
```
Comment (luka-ethernal) [inline kate/recovery/src/com.rs:89]: This is obvious what it does, but a comment couldn't hurt.

**Review by luka-ethernal — APPROVED** [review]

File: kate/recovery/Cargo.toml [hunk 5]
```diff
@@ -5,6 +5,7 @@ authors = ["Denis Ermolin <denis.ermolin@matic.network>"]
 edition = "2018"
 
 [dependencies]
+anyhow = "1.0.41"
```
[the remark this answers is not in the document]
Comment (aterentic-ethernal) [reply kate/recovery/Cargo.toml]: This is done, I have added enum with errors.

File: kate/recovery/src/com.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (aterentic-ethernal) [reply kate/recovery/src/com.rs:12]: @MiguelDD1 , Ill leave it as it is and we can discuss in a followong weeks how to improve it and remove duplication.

File: kate/recovery/src/com.rs (hunk 3 above)
Comment (aterentic-ethernal) [reply kate/recovery/src/com.rs:31]: Done.

File: kate/recovery/src/com.rs (hunk 4 above)
Comment (aterentic-ethernal) [reply kate/recovery/src/com.rs:131]: Done.

**Review by kroos47 — APPROVED** [review]

**Review by luka-ethernal — APPROVED** [review]

**Pull request merged**